### ISSUE TITLE
Unchecked method invocation / conversion compile warnings

### ISSUE THREAD

**Issue by svenjacobs** [google/dagger#199]
I have the unchecked lint compiler option enabled in my project (via `-Xlint:unchecked`) and get a few `unchecked method invocation` and `unchecked conversion` warnings in the generated components. For instance:

```
/Users/sven/Dev/Projects/01_Private/leettime/build/generated/source/apt/release/com/svenjacobs/leettime/app/ui/DaggerMainActivityComponent.java:57: warning: [unchecked] unchecked method invocation: method create in class MainActivity_MembersInjector is applied to given types
    this.mainActivityMembersInjector = MainActivity_MembersInjector.create((MembersInjector) MembersInjectors.noOp(), leetAlarmManagerProvider, leetNotificationControllerProvider);
                                                                          ^
  required: MembersInjector<ButterKnifeActivity>,Provider<LeetAlarmManager>,Provider<LeetNotificationController>
  found: MembersInjector,Provider<LeetAlarmManager>,Provider<LeetNotificationController>
/Users/sven/Dev/Projects/01_Private/leettime/build/generated/source/apt/release/com/svenjacobs/leettime/app/ui/DaggerMainActivityComponent.java:57: warning: [unchecked] unchecked conversion
    this.mainActivityMembersInjector = MainActivity_MembersInjector.create((MembersInjector) MembersInjectors.noOp(), leetAlarmManagerProvider, leetNotificationControllerProvider);
                                                                           ^
  required: MembersInjector<ButterKnifeActivity>
  found:    MembersInjector
/Users/sven/Dev/Projects/01_Private/leettime/build/generated/source/apt/release/com/svenjacobs/leettime/app/ui/DaggerMainActivityComponent.java:57: warning: [unchecked] unchecked conversion
    this.mainActivityMembersInjector = MainActivity_MembersInjector.create((MembersInjector) MembersInjectors.noOp(), leetAlarmManagerProvider, leetNotificationControllerProvider);
                                                                          ^
```

This seems to be always related to the invocation of `MembersInjector.noOp()`, where the type argument is missing. For example

```java
this.mainActivityMembersInjector = MainActivity_MembersInjector.create((MembersInjector) MembersInjectors.noOp(), leetAlarmManagerProvider, leetNotificationControllerProvider);

```

should actually be generated as

```java
this.mainActivityMembersInjector = MainActivity_MembersInjector.create(MembersInjectors.<ButterKnifeActivity>noOp(), leetAlarmManagerProvider, leetNotificationControllerProvider);
```

in this case. This is on 2.0.1.

**Comment by gk5885** [comment]
While we'd _like_ to be able to generate the code you suggested, this is actually an elaborate way to dodge visibility restrictions.  If `ButterKnifeActivity` is package-private and in a different package from your component, the alternative won't compile.

That said, we should definitely suppress the warning since we are clearly doing it on purpose.

**Comment by svenjacobs** [comment]
Oh, I see. I didn't have possible visibility restrictions in mind. In this case suppressing the warning is perfectly fine :)

**Comment by tbroyer** [comment]
Unless I'm mistaken, those "raw types casts" only happens in the `initialize` methods, so suppressing warnings on those methods unconditionally (whether they'll contain such casts or not) would probably work.

**Comment by tavianator** [comment]
@tbroyer But then you will get an `Unnecessary @SuppressWarnings("rawtypes")` warning.

**Comment by svenjacobs** [comment]
Couldn't you check during code generation of the component if the class is visible to the component and if so apply the generic type arguments as proposed by me, else apply `@SuppressWarning`?

**Comment by gk5885** [comment]
Indeed you could, @svenjacobs.  That's probably what the fix will look like.

**Comment by gildor** [comment]
+1

**Comment by rusmonster** [comment]
+1

**Comment by BenGroot** [comment]
+1 any news on this?

**Comment by MatthijsKamstra** [comment]
:+1:

**Comment by jpink** [comment]
+1

**Comment by netdpb** [comment]
This was fixed by 28032ee17a8207d76f68d855aebde82d1a5aeb3d.
